REGRESSION (289928@main): Svelte transition fade + transition scale results in flashing

nipunshukla21 opened the pull request:
#### 42b3f8d9178498cd6a992ee304cdaf0bf215f936
<pre>
REGRESSION (289928@main): Svelte transition fade + transition scale results in flashing
<a href="https://bugs.webkit.org/show_bug.cgi?id=303217">https://bugs.webkit.org/show_bug.cgi?id=303217</a>
<a href="https://rdar.apple.com/165583864">rdar://165583864</a>

Reviewed by Antoine Quint.

289928@main centralized effect stack membership updates but inadvertently
removed immediate stopping of accelerated animations when effects are removed
from the stack. This patch restores that, preventing a visible flash. It also
ensures that when animations are removed, they are allowed to finish before
removal.

* LayoutTests/webanimations/simultaneous-animations-removed-separately-expected.txt: Added.
* LayoutTests/webanimations/simultaneous-animations-removed-separately.html: Added.
* Source/WebCore/animation/KeyframeEffect.cpp:
(WebCore::KeyframeEffect::wasRemovedFromEffectStack):
* Source/WebCore/animation/WebAnimation.h:
(WebCore::WebAnimation::hasPendingFinishNotification const):

Canonical link: <a href="https://commits.webkit.org/304710@main">https://commits.webkit.org/304710@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/078f2a0f714a81f465ba6fdea78ceff4743188c0

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/136310 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/8667 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/47590 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/144022 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/89281 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/138181 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/9347 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/8511 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/104226 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/89281 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/139255 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/6805 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/122144 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/85059 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/6458 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/4120 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/4613 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/115749 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/40345 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/146766 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/8349 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/40913 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/112568 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/8366 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/7017 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/112912 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/6389 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/118449 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/62336 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/21014 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/8397 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/36506 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/8115 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/71956 "Built successfully") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/8337 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/8189 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

graouts commented on Source/WebCore/animation/KeyframeEffect.cpp:
```diff
@@ -2238,6 +2239,36 @@ void KeyframeEffect::wasAddedToEffectStack()
+
+    // If the effect was running accelerated, we need to mark it for removal straight away
+    // since it will not be invalidated by a future call to KeyframeEffectStack::applyPendingAcceleratedActions().
+    if (animation() && (isRunningAccelerated() || isAboutToRunAccelerated())) {
```
If `canBeAccelerated()` is `true` then you're guaranteed to have an associated animation here. So I would remove the `animation()` check and ASSERT that it exists instead.

graouts commented on Source/WebCore/animation/KeyframeEffect.cpp:
```diff
@@ -2238,6 +2239,36 @@ void KeyframeEffect::wasAddedToEffectStack()
+    // If the effect was running accelerated, we need to mark it for removal straight away
+    // since it will not be invalidated by a future call to KeyframeEffectStack::applyPendingAcceleratedActions().
+    if (animation() && (isRunningAccelerated() || isAboutToRunAccelerated())) {
+        auto* anim = animation();
```
I'm surprised the safer-cpp bot didn't complain about this not being ref-counted since `WebAnimation::playState()` is not a trivial function. We typically don't shorten names like `anim` instead of `animation`.
```suggestion
        Ref animation = *animation();
```

nipunshukla21 committed:
REGRESSION (289928@main): Svelte transition fade + transition scale results in flashing
https://bugs.webkit.org/show_bug.cgi?id=303217
rdar://165583864

Reviewed by Antoine Quint.

289928@main centralized effect stack membership updates but inadvertently
removed immediate stopping of accelerated animations when effects are removed
from the stack. This patch restores that, preventing a visible flash. It also
ensures that when animations are removed, they are allowed to finish before
removal.

* LayoutTests/webanimations/simultaneous-animations-removed-separately-expected.txt: Added.
* LayoutTests/webanimations/simultaneous-animations-removed-separately.html: Added.
* Source/WebCore/animation/KeyframeEffect.cpp:
(WebCore::KeyframeEffect::wasRemovedFromEffectStack):
* Source/WebCore/animation/WebAnimation.h:
(WebCore::WebAnimation::hasPendingFinishNotification const):

Canonical link: https://commits.webkit.org/304710@main
